### DISCUSSION TITLE
SetOwningMol

### DISCUSSION THREAD

**Discussion by woodg07** [rdkit/rdkit#3766] [Q&A] [answered]
Hello,

I have a couple of function definitions in a wider class that I'm writing in python.  For one function I embed the molecules with n number of conformers, I then want to make sure that these conformers remain with the parsed molecule after the function call. I thought this was done by reference under the hood so I wouldn't have to specifically set the conformers (see code snippet 1. below). However, this doesn't see to be the case.  I tried to parse back the conformers and then do a setOwningMol but I get an error saying that the conformer class has no attribute setOwningMol or SetOwningMol (see snippet 2.) Any ideas on how to proceed would be great.

snippet 1:

    gen_conformers(rdkitmol,n)
    rdkmol.GetNumConformers() #### returns 0 here, I thought this would be done by reference?

    def gen_conformers(rdkitmol,n):
        cids = AllChem.EmbedMultipleConfs(rdkmol,numConfs=n)
        rdkmol.GetNumConformers() #### returns n here

snippet 2:

    confs = gen_conformers(rdkitmol,n)
    for conf in confs:
        conf.SetOwningMol(rdkitmol)  #### 'Conformer' object has no attribute 'SetOwningMol' I tried with setOwningMol also

    def gen_conformers(rdkitmol,n):
        cids = AllChem.EmbedMultipleConfs(rdkmol,numConfs=n)
        return rdkmol.GetConformers()

**Comment by bp-kelley** [comment]
The function you are looking for is AddConformer

```
>>> help(mol.AddConformer)
Help on method AddConformer:

AddConformer(...) method of rdkit.Chem.rdchem.Mol instance
    AddConformer( (Mol)self, (Conformer)conf [, (bool)assignId=False]) -> int :
        Add a conformer to the molecule and return the conformer ID
    
        C++ signature :
            unsigned int AddConformer(RDKit::ROMol {lvalue},RDKit::Conformer* [,bool=False])
```

**Comment by greglandrum** [comment, marked as answer]
@woodg07 I don't really understand what you're trying to do.

`AllChem.EmbedMultipleConfs` adds conformations to a molecule; you don't need to do any additional work to add the conformers to the molecule:
```
>>> from rdkit.Chem import AllChem
>>> from rdkit import Chem
>>> m = Chem.AddHs(Chem.MolFromSmiles('CCCO'))
>>> cids = AllChem.EmbedMultipleConfs(m,5)
>>> m.GetNumConformers()
5
```

If your concern is that adding the conformers in a function may not change the molecule in the calling scope, you don't need to be:
```
>>> def gen_confs(mol,n):
...     AllChem.EmbedMultipleConfs(mol,n)
... 
>>> m = Chem.AddHs(Chem.MolFromSmiles('CCCO'))
>>> gen_confs(m,7)
>>> m.GetNumConformers()
7
```

Does that answer the question? If not, can you please clarify what you're trying to do?

**Reply by woodg07** [reply]
Thank you for following up. I was having an issue with passing a copy of the molecule to the function so it looked like it wasn't updating. I appreciate that you wanted more details but I think we can close the problem now. Thank you
